src/tacacs_plus_fmt_plug.c

```diff
@@ -0,0 +1,337 @@
+		goto err;
+	if ((p = strtokm(NULL, "$")) == NULL)   // hash_data
+		goto err;
+	if (hexlenl(p, &extra) != 2 * 2 || extra)
+		goto err;
+
+
+	MEM_FREE(keeptr);
+	return 1;
+
+err:
+	MEM_FREE(keeptr);
+	return 0;
+}
+
+static void *get_salt(char *ciphertext)
+{
+	static struct custom_salt cs;
+	char *ctcopy = strdup(ciphertext);
+	char *keeptr = ctcopy;
+	char *p;
+	int i;
+
+	memset(&cs, 0, SALT_SIZE);
+	ctcopy += TAG_LENGTH;
+	p = strtokm(ctcopy, "$");
+	p = strtokm(NULL, "$");
+	cs.pre_hash_data_len = strlen(p) / 2;
+	for (i = 0; i < cs.pre_hash_data_len; i++)
+		cs.pre_hash_data[i] = (atoi16[ARCH_INDEX(p[2 * i])] << 4) | atoi16[ARCH_INDEX(p[2 * i + 1])];
+	p = strtokm(NULL, "$");
+	cs.ctlen = strlen(p) / 2;
+	for (i = 0; i < cs.ctlen; i++)
+		cs.ciphertext[i] = (atoi16[ARCH_INDEX(p[2 * i])] << 4) | atoi16[ARCH_INDEX(p[2 * i + 1])];
+	p = strtokm(NULL, "$");
+	cs.hash_data_len = 2;
+	for (i = 0; i < 2; i++)
+		cs.hash_data[i] = (atoi16[ARCH_INDEX(p[2 * i])] << 4) | atoi16[ARCH_INDEX(p[2 * i + 1])];
+	MD5_Init(&cs.pctx);
+	MD5_Update(&cs.pctx, cs.pre_hash_data, cs.pre_hash_data_len);
+
+	MEM_FREE(keeptr);
+
+	return &cs;
+}
+
+static int check_password(char *secret, struct custom_salt *cs)
+{
+	MD5_CTX cur;
+	unsigned char digest[MD5_DIGEST_LENGTH];
+	int i;
+	unsigned char status, flags;
+	unsigned short server_msg_len, data_len;
+	unsigned char cleartext[MD5_DIGEST_LENGTH];
+	int length;
+
+	length = (cs->ctlen < MD5_DIGEST_LENGTH ? cs->ctlen : MD5_DIGEST_LENGTH);
+	memcpy((void *)&cur, &cur_salt->pctx, sizeof(MD5_CTX));
+	MD5_Update(&cur, secret, strlen(secret));
+	MD5_Update(&cur, cs->hash_data, cs->hash_data_len);
+	MD5_Final(digest, &cur);
+
+	for (i = 0; i < length; i++) {
+		cleartext[i] = cs->ciphertext[i] ^ digest[i];
+	}
+
+	status = cleartext[0];
+	flags = cleartext[1];
+	server_msg_len = cleartext[3] | (cleartext[2] << 8);
+	data_len = cleartext[5] | (cleartext[4] << 8);
+	if(((status >= 0x01 && status <= 0x07) || status == 0x21) && (flags == 0x01 || flags == 0x00) &&
+		(6 + server_msg_len + data_len == cs->ctlen)) {
+		return 1;
+	}
+
+	return 0;
+}
+
+static void set_salt(void *salt)
+{
+	cur_salt = (struct custom_salt *)salt;
+}
+
+static int crypt_all(int *pcount, struct db_salt *salt)
+{
+	const int count = *pcount;
+	int index = 0;
+
+	if (any_cracked) {
+		memset(cracked, 0, cracked_size);
+		any_cracked = 0;
+	}
+#ifdef _OPENMP
+#pragma omp parallel for
+#endif
+	for (index = 0; index < count; index++)
+	{
+		if (check_password(saved_key[index], cur_salt)) {
+			cracked[index] = 1;
+#ifdef _OPENMP
+#pragma omp atomic
+#endif
+			any_cracked |= 1;
+		}
+	}
+
+	return count;
+}
+
+static int cmp_all(void *binary, int count)
+{
+	return any_cracked;
+}
+
+static int cmp_one(void *binary, int index)
+{
+	return cracked[index];
+}
+
+static int cmp_exact(char *source, int index)
+{
+	return cracked[index];
+}
+
+static void tacacsplus_set_key(char *key, int index)
+{
+	strnzcpy(saved_key[index], key, PLAINTEXT_LENGTH + 1);
```

Comment:
Could probably pre-compute the length here and use it instead of the `strlen(secret)` during hashing.